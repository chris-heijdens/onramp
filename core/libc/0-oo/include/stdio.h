/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2023-2024 Fraser Heavy Software
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#ifndef __ONRAMP_LIBC_STDIO_H_INCLUDED
#define __ONRAMP_LIBC_STDIO_H_INCLUDED

#ifndef __onramp_libc__
    #error "__onramp/__predef.h must be force-included by the preprocessor before any libc headers."
#endif

#include <__onramp/__size_t.h>
#include <__onramp/__null.h>

#define EOF (-1)

typedef void FILE;

extern FILE* stdin;
extern FILE* stdout;
extern FILE* stderr;

FILE* fopen(const char* filename, const char* mode);
size_t fread(void* ptr, size_t size, size_t nmemb, FILE* file);
size_t fwrite(const void* ptr, size_t size, size_t nmemb, FILE* file);
int fclose(FILE* file);
int feof(FILE* file);

long ftell(FILE* file);
int fseek(FILE* stream, long offset, int base);

#define SEEK_SET 0
#define SEEK_CUR 1
#define SEEK_END 2

int fputc(int c, FILE* file);
int fputs(const char* s, FILE* file);
int putchar(int c);
int puts(const char* s);

/* TODO these are probably not going to be defined by libc/0, only by libc/2 or
 * libc/3
typedef int fpos_t;
int fgetpos(FILE* file, fpos_t* pos);
int fsetpos(FILE* file, const fpos_t* pos);
*/

// TODO only defined by libc/1
int remove(const char* filename);

// posix
int fileno(FILE* file);
typedef int mode_t;
int fchmod(int fileno, int mode); // TODO this belongs in sys/stat.h

#endif
