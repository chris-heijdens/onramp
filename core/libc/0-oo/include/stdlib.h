/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2023-2024 Fraser Heavy Software
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#ifndef __ONRAMP_LIBC_STDLIB_H_INCLUDED
#define __ONRAMP_LIBC_STDLIB_H_INCLUDED

#ifndef __onramp_libc__
    #error "__onramp/__predef.h must be force-included by the preprocessor before any libc headers."
#endif

#include <__onramp/__size_t.h>
#include <__onramp/__null.h>

#define EXIT_SUCCESS 0
#define EXIT_FAILURE 1

void* malloc(size_t size);
void* calloc(size_t element_count, size_t element_size);
void free(void* p);
void* realloc(void* old_ptr, size_t new_size);

void _Exit(int status);

/**
 * Returns the bounds of the largest region of memory that is currently
 * unallocated.
 *
 * This can be used as the heap for child processes.
 *
 * The memory is not actually allocated. You must not free() this.
 *
 * TODO move this to an internal onramp header
 */
void* __malloc_largest_unused_region(size_t* out_size);

/**
 * Gets the value of the environment variable with the given name.
 */
char* getenv(const char* key);

#endif
