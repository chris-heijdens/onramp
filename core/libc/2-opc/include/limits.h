/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2023-2024 Fraser Heavy Software
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#ifndef __ONRAMP_LIBC_LIMITS_H_INCLUDED
#define __ONRAMP_LIBC_LIMITS_H_INCLUDED

#ifndef __onramp_libc__
    #error "__onramp/__predef.h must be force-included by the preprocessor before any libc headers."
#endif

#define CHAR_BIT 8
#define CHAR_MAX SCHAR_MAX
#define CHAR_MIN SCHAR_MIN
#define SCHAR_MAX 127
#define SCHAR_MIN (-127-1)
#define UCHAR_MAX 255U

#define SHRT_MAX 32767
#define SHRT_MIN (-32767-1)
#define USHRT_MAX 65535U

#define INT_MAX 2147483647
#define INT_MIN (-2147483647-1)
#define UINT_MAX 4294967295U

#define LONG_MAX INT_MAX
#define LONG_MIN INT_MIN
#define ULONG_MAX UINT_MAX

#define LLONG_MAX LONG_MAX
#define LLONG_MIN LONG_MIN
#define ULLONG_MAX ULONG_MAX

#define MB_LEN_MAX 1

#endif
