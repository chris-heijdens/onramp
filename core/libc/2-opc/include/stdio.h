/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2023-2024 Fraser Heavy Software
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#ifndef __ONRAMP_LIBC_STDIO_H_INCLUDED
#define __ONRAMP_LIBC_STDIO_H_INCLUDED

#ifndef __onramp_libc__
    #error "__onramp/__predef.h must be force-included by the preprocessor before any libc headers."
#endif

#include <__onramp/__size_t.h>
#include <__onramp/__null.h>
#include <__onramp/__va_list.h>

struct __file;
typedef struct __file FILE;
typedef unsigned int fpos_t;

#define _IOFBF 1
#define _IOLBF 2
#define _IONBF 3

#define BUFSIZ 128
#define EOF (-1)
#define FOPEN_MAX 13 // TODO
#define FILENAME_MAX 256
#define L_tmpnam 256 // TODO
#define TMP_MAX 1000 // TODO

#define SEEK_CUR 1
#define SEEK_END 2
#define SEEK_SET 3

extern FILE* __stdin;
extern FILE* __stdout;
extern FILE* __stderr;
#define stdin (__stdin)
#define stdout (__stdout)
#define stderr (__stderr)

// standard C

int remove(const char* filename);
int rename(const char* old, const char* new);
FILE* tmpfile(void);
char* tmpnam(char* s);
int fclose(FILE* stream);
int fflush(FILE* stream);
FILE* fopen(const char* restrict filename, const char* restrict mode);
FILE* freopen(const char* restrict filename, const char* restrict mode, FILE* restrict stream);
int setvbuf(FILE* restrict stream, char* restrict buf, int mode, size_t size);

int fprintf(FILE* restrict stream, const char* restrict format, ...);
int fscanf(FILE* restrict stream, const char* restrict format, ...);
int printf(const char* restrict format, ...);
int scanf(const char* restrict format, ...);
int snprintf(char* restrict buffer, size_t n, const char* restrict format, ...);
int sprintf(char* restrict buffer, const char* restrict format, ...);
int sscanf(const char* restrict buffer, const char* restrict format, ...);
int vfprintf(FILE* restrict stream, const char* restrict format, va_list arg);
int vfscanf(FILE* restrict stream, const char* restrict format, va_list arg);
int vprintf(const char* restrict format, va_list arg);
int vscanf(const char* restrict format, va_list arg);
int vsnprintf(char* restrict buffer, size_t n, const char* restrict format, va_list arg);
int vsprintf(char* restrict buffer, const char* restrict format, va_list arg);
int vsscanf(const char* restrict buffer, const char* restrict format, va_list arg);

//[[deprecated]] char* gets(char* s);
int fgetc(FILE* stream);
char* fgets(char* restrict s, int n, FILE* restrict stream);
int fputc(int c, FILE* stream);
int fputs(const char* restrict s, FILE* restrict stream);
int getc(FILE* stream);
int getchar(void);
int putc(int c, FILE* stream);
int putchar(int c);
int puts(const char* s);
int ungetc(int c, FILE* stream);

size_t fread(void* restrict ptr, size_t size, size_t nmemb, FILE* restrict stream);
size_t fwrite(const void* restrict ptr, size_t size, size_t nmemb, FILE* restrict stream);
int fgetpos(FILE* restrict stream, fpos_t*  restrict pos);
int fseek(FILE* stream, long offset, int whence);
int fsetpos(FILE* stream, const fpos_t* pos);
long ftell(FILE* stream);
void rewind(FILE* stream);

void clearerr(FILE* stream);
int feof(FILE* stream);
int ferror(FILE* stream);
void perror(const char* s);

// gnu extensions
#ifdef _GNU_SOURCE
int asprintf(char** restrict out_string, const char* restrict format, ...);
int vasprintf(char** restrict out_string, const char* restrict format, va_list args);
#endif

#endif
