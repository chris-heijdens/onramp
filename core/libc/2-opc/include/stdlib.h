/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2023-2024 Fraser Heavy Software
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#ifndef __ONRAMP_LIBC_STDLIB_H_INCLUDED
#define __ONRAMP_LIBC_STDLIB_H_INCLUDED

#ifndef __onramp_libc__
    #error "__onramp/__predef.h must be force-included by the preprocessor before any libc headers."
#endif

#include <__onramp/__size_t.h>
#include <__onramp/__wchar_t.h>
#include <__onramp/__null.h>

#define EXIT_SUCCESS 0
#define EXIT_FAILURE 1

int atoi(const char* nptr);
long int atol(const char* nptr);

long int strtol(const char* restrict nptr, char** restrict endptr, int base);
unsigned long int strtoul(const char* restrict nptr, char** restrict endptr, int base);

_Noreturn void abort(void);
_Noreturn void _Exit(int status);
char* getenv(const char* name);
int system(const char* string);

void* aligned_alloc(size_t __alignment, size_t __size);
void* calloc(size_t __count, size_t __element_size);
void free(void* __ptr);
void* malloc(size_t __size);
void* realloc(void* __ptr, size_t __size);
int posix_memalign(void** __ptr, size_t __alignment, size_t __size);
// TODO extensions
size_t malloc_size(void* ptr);

#endif
