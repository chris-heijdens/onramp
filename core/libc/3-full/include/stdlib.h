/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2023-2024 Fraser Heavy Software
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#ifndef __ONRAMP_LIBC_STDLIB_H_INCLUDED
#define __ONRAMP_LIBC_STDLIB_H_INCLUDED

#ifndef __onramp_libc__
    #error "__onramp/__predef.h must be force-included by the preprocessor before any libc headers."
#endif

#include <__onramp/__size_t.h>
#include <__onramp/__wchar_t.h>
#include <__onramp/__null.h>
#include <__onramp/__malloc.h>

#define EXIT_SUCCESS 0
#define EXIT_FAILURE 1

// TODO MB_CUR_MAX

double atof(const char* nptr);

int atoi(const char* nptr);
long int atol(const char* nptr);
long long int atoll(const char* nptr);

double strtod(const char* restrict nptr, char** restrict endptr);
float strtof(const char* restrict nptr, char** restrict endptr);
long double strtold(const char* restrict nptr, char** restrict endptr);

long int strtol(const char* restrict nptr, char** restrict endptr, int base);
long long int strtoll(const char* restrict nptr, char** restrict endptr, int base);
unsigned long int strtoul(const char* restrict nptr, char** restrict endptr, int base);
unsigned long long int strtoull(const char* restrict nptr, char** restrict endptr, int base);

#define RAND_MAX 65535
int rand(void);
void srand(unsigned int seed);
//#if 1 // TODO POSIX 199506L
int rand_r(unsigned* seed);
//#endif

_Noreturn void abort(void);
int atexit(void (*func)(void));
int at_quick_exit(void (*func)(void));
_Noreturn void exit(int status);
_Noreturn void _Exit(int status);
char* getenv(const char* name);
_Noreturn void quick_exit(int status);
int system(const char* string);

void* aligned_alloc(size_t __alignment, size_t __size);
void* calloc(size_t __count, size_t __element_size);
void free(void* __ptr);
void* malloc(size_t __size);
void* realloc(void* __ptr, size_t __size);
int posix_memalign(void** __ptr, size_t __alignment, size_t __size);
// TODO extensions
size_t malloc_size(void* ptr);

// Searching and sorting utilities
void *bsearch(const void* __key, const void* __base,
        size_t __element_count, size_t __element_size,
        int (*__user_compare)(const void* __key, const void* __other));
void qsort(void* __first, size_t __element_count, size_t __element_size,
        int (*__user_compare)(const void* __left, const void* __right));
void qsort_r(void* __first, size_t __element_count, size_t __element_size,
        int (*__user_compare)(const void* __left, const void* __right, void* __user_context),
        void* __user_context);

// Integer arithmetic functions
int abs(int j);
long int labs(long int j);
long long int llabs(long long int j);
//div_t div(int numer, int denom);
//ldiv_t ldiv(long int numer, long int denom);
//lldiv_t lldiv(long long int numer, long long int denom

/* Multibyte/wide character conversion functions */
// TODO

#endif
