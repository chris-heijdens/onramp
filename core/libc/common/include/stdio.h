/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2023-2024 Fraser Heavy Software
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#ifndef __ONRAMP_LIBC_STDIO_H_INCLUDED
#define __ONRAMP_LIBC_STDIO_H_INCLUDED

#ifndef __onramp_libc__
    #error "__onramp/__predef.h must be force-included by the preprocessor before any libc headers."
#endif

#include <__onramp/__size_t.h>
#include <__onramp/__null.h>
#include <__onramp/__va_list.h>



// FILE is an incomplete struct so that it's not convertible to anything else.
// (omC doesn't have structs so it's just void when bootstrapping cci/1.)

#ifdef __onramp_cci_omc__
typedef void FILE;
#endif

#ifndef __onramp_cci_omc__
struct __file;
typedef struct __file FILE;
#endif

// These are supposed to be macros. They also need to be real variables to be
// accessible from assembly and backwards-compatible with previous stages, but
// we can't define recursive macros with the omC preprocessor. We therefore
// use this workaround to define them.
extern FILE* stdin;
extern FILE* stdout;
extern FILE* stderr;
#ifndef __ONRAMP_LIBC_2_FILE_IMPL
    #ifndef __onramp_cpp_omc__
        #define stdin ((FILE*)stdin)
        #define stdout ((FILE*)stdout)
        #define stderr ((FILE*)stderr)
    #endif
#endif



typedef long fpos_t;

#define _IOFBF 1
#define _IOLBF 2
#define _IONBF 3

#define BUFSIZ 512
#define EOF (-1)
#define FOPEN_MAX 13 // TODO
#define FILENAME_MAX 256
#define L_tmpnam 256 // TODO
#define TMP_MAX 1000 // TODO

#define SEEK_SET 0
#define SEEK_CUR 1
#define SEEK_END 2



// omC

FILE* fopen(const char* restrict filename, const char* restrict mode);
size_t fread(void* restrict ptr, size_t element_size, size_t element_count, FILE* restrict file);
size_t fwrite(const void* restrict ptr, size_t element_size, size_t element_count, FILE* restrict file);
int fclose(FILE* file);
int feof(FILE* file);

int fgetc(FILE* file);
int fputc(int c, FILE* file);
int fputs(const char* restrict s, FILE* restrict file);
int putchar(int c);
int putc(int c, FILE* file);
int puts(const char* s);

int fseek(FILE* file, long offset, int whence);
long ftell(FILE* file);

int remove(const char* filename);

int chmod(const char* filename, int mode); // TODO this belongs in sys/stat.h



// opC

#ifndef __onramp_cci_omc__

int rename(const char* old, const char* new);
FILE* tmpfile(void);
char* tmpnam(char* s);
int fflush(FILE* file);
FILE* freopen(const char* restrict filename, const char* restrict mode, FILE* restrict file);
int setvbuf(FILE* restrict file, char* restrict buf, int mode, size_t size);

int fprintf(FILE* restrict file, const char* restrict format, ...);
int fscanf(FILE* restrict file, const char* restrict format, ...);
int printf(const char* restrict format, ...);
int scanf(const char* restrict format, ...);
int snprintf(char* restrict buffer, size_t n, const char* restrict format, ...);
int sprintf(char* restrict buffer, const char* restrict format, ...);
int sscanf(const char* restrict buffer, const char* restrict format, ...);
int vfprintf(FILE* restrict file, const char* restrict format, va_list arg);
int vfscanf(FILE* restrict file, const char* restrict format, va_list arg);
int vprintf(const char* restrict format, va_list arg);
int vscanf(const char* restrict format, va_list arg);
int vsnprintf(char* restrict buffer, size_t n, const char* restrict format, va_list arg);
int vsprintf(char* restrict buffer, const char* restrict format, va_list arg);
int vsscanf(const char* restrict buffer, const char* restrict format, va_list arg);

//[[deprecated]] char* gets(char* s);
char* fgets(char* restrict s, int n, FILE* restrict file);
int getc(FILE* file);
int getchar(void);
int ungetc(int c, FILE* file);

int fgetpos(FILE* restrict file, fpos_t* restrict pos);
int fsetpos(FILE* file, const fpos_t* pos);
void rewind(FILE* file);

void clearerr(FILE* file);
int ferror(FILE* file);
void perror(const char* s);

// posix
FILE* fdopen(int fd, const char* mode);

// gnu extensions
#ifdef _GNU_SOURCE
int asprintf(char** restrict out_string, const char* restrict format, ...);
int vasprintf(char** restrict out_string, const char* restrict format, va_list args);
#endif

// posix
int fileno(FILE* file);
typedef int mode_t;

#endif



#endif
