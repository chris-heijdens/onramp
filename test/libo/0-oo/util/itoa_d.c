#include <stdlib.h>
#include <stdio.h>
#include <limits.h>

void __itoa_d(int value, char* buffer);

char* buf;

int main(void) {
    buf = malloc(12);

    __itoa_d(0, buf);
    puts(buf);

    __itoa_d(1, buf);
    puts(buf);

    __itoa_d(9, buf);
    puts(buf);

    __itoa_d(15, buf);
    puts(buf);

    __itoa_d(1234567, buf);
    puts(buf);

    __itoa_d(INT_MAX, buf);
    puts(buf);

    __itoa_d(-1, buf);
    puts(buf);

    __itoa_d(-2, buf);
    puts(buf);

    __itoa_d(-9, buf);
    puts(buf);

    __itoa_d(-99, buf);
    puts(buf);

    __itoa_d(-1234567, buf);
    puts(buf);

    __itoa_d(INT_MIN, buf);
    puts(buf);
}
