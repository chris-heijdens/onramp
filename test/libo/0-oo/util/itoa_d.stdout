0
1
9
15
1234567
2147483647
-1
-2
-9
-99
-1234567
-2147483648
